lib/ecto/type.ex

```diff
@@ -239,51 +251,51 @@ defmodule Ecto.Type do
     end
   end
 
-  @doc """
-  Dumps the `value` for `type` considering it will be embedded in `format`.
+  # @doc """
+  # Dumps the `value` for `type` considering it will be embedded in `format`.
 
-  ## Examples
+  # ## Examples
 
-      iex> Ecto.Type.embedded_dump(:decimal, Decimal.new("1"), :json)
-      {:ok, Decimal.new("1")}
+  #     iex> Ecto.Type.embedded_dump(:decimal, Decimal.new("1"), :json)
+  #     {:ok, Decimal.new("1")}
 
-  """
-  def embedded_dump({:embed, _} = type, value, format) do
-    dump(type, value, &embedded_dump(&1, &2, format))
-  end
+  # """
+  # def embedded_dump({:embed, _} = type, value, format) do
+  #   dump(type, value, &embedded_dump(&1, &2, format))
+  # end
 
-  def embedded_dump(type, value, format) do
-    case embed_as(type, format) do
-      :self -> {:ok, value}
-      :dump -> dump(type, value)
-    end
-  end
+  # def embedded_dump(type, value, format) do
+  #   case embed_as(type, format) do
+  #     :self -> {:ok, value}
+  #     :dump -> dump(type, value)
+  #   end
+  # end
 
-  @doc """
-  Loads the `value` for `type` considering it was embedded in `format`.
+  # @doc """
+  # Loads the `value` for `type` considering it was embedded in `format`.
 
-  ## Examples
+  # ## Examples
 
-      iex> Ecto.Type.embedded_load(:decimal, "1", :json)
-      {:ok, Decimal.new("1")}
+  #     iex> Ecto.Type.embedded_load(:decimal, "1", :json)
+  #     {:ok, Decimal.new("1")}
 
-  """
-  def embedded_load({:embed, _} = type, value, format) do
-    load(type, value, &embedded_load(&1, &2, format))
-  end
+  # """
+  # def embedded_load({:embed, _} = type, value, format) do
+  #   load(type, value, &embedded_load(&1, &2, format))
+  # end
 
-  def embedded_load(type, value, format) do
-    case embed_as(type, format) do
-      :self ->
-        case cast(type, value) do
-          {:ok, _} = ok -> ok
-          _ -> :error
-        end
+  # def embedded_load(type, value, format) do
+  #   case embed_as(type, format) do
+  #     :self ->
+  #       case cast(type, value, nil) do
+  #         {:ok, _} = ok -> ok
+  #         _ -> :error
+  #       end
 
-      :dump ->
-        load(type, value)
-    end
-  end
+  #     :dump ->
+  #       load(type, value)
+  #   end
+  # end
```

Comment:
That's a breaking change. However I think those parts with pattern-matching like `{:embed, _}` could be changed to `{:parameterized, _, _}` or removed if no longer needed as `{:embed, _}` match is not documented.

Reply:
Yes, the idea is to remove "embed" as hardcoded logic into Embed, and this function is pure {:embed, _} logic, hence the removal.

Reply:
@TheFirstAvenger Pure `{:embed, _}` logic is still in public API which means developers working on libraries/projects may use this until it's deprecated. We can't remove public API "just like that", because it's not needed for new projects.

This is why we also support `embeds_many/3` and `embeds_one/3`. Making breaking change would stop developers from updating dependencies which may cause critical problems especially in case next releases would have bug fixes and/or security fixes or (in best case) other dependencies would require newer versions of `ecto`.

Therefore even if it's not important for new projects please write a backwards-compatible API for `{:parameterized, _, _}`. Of course supporting `{:embed, _}` should no be important as developers should just fetch field types dynamically from `schema` definition.